### ISSUE TITLE
PipeReader.OnWriterCompleted callback exception will crash the process

### ISSUE THREAD

**Issue by AArnott** [dotnet/corefx#31695]
Apparently `PipeReader.OnWriterCompleted` (and perhaps `PipeWriter.OnReaderCompleted`?) execute their callbacks directly on the threadpool, with no `Task` or other means of wrapping exceptions thrown. As a result, if these callbacks throw an exception, the process crashes.

I find this very unexpected, since it is not clear how the callback will be invoked, and it's not documented to be directly on the threadpool. Most callback systems do not crash the process when the callback throws.

**Comment by garethwarry** [comment]
Pipelines in general has been a really awful experience... no documentation, tons of bugs... 
Not production ready in the slightest.

**Comment by davidfowl** [comment]
@AArnott I wasn’t aware that things with BCL libraries with callbacks don’t crash the process when user code throws fail. I need to take a look and see if that’s indeed true (I can even think of a couple of cases where that isn’t true). We can look at adding a try/catch if need be. There really isn’t really a general way to handle this AFAIK. We’re not using Task dispatch because of the in unnecessary overhead.

@garethwarry the doc situation is a bit unfortunate. We’re working on getting basic API docs out. I’m sorry you’ve had such a poor experience with them but I can assure you they are “production ready” as they are used in ASP.NET Core itself. If you have bugs or issues with the user experience then please be sure to file them so they can be fixed.

**Comment by garethwarry** [comment]
@davidfowl Well, I decided to implement my own duplex pipe for sockets, but I'll gladly give feedback on why my implementation with System.IO.Pipelines failed. I'll make a separate issue to track that.

**Comment by davidfowl** [comment]
@garethwarry sounds good.

**Comment by AArnott** [comment]
> We’re not using Task dispatch because of the in unnecessary overhead.

I think the unnecessary is arguable here. It removes from the consumer the option of how their callback should be scheduled, making it difficult or impossible to track exactly when their callbacks have completed firing (so, for example, they could ensure they aren't executing during a shutdown). Customers have become quite comfortable with `Task` and how it behaves. IMHO this callback API felt like a throwback to .NET 3.5 days. I get that you're trying to reduce overhead, but couldn't a `ValueTask Completion` property exposed on the pipe API been efficient enough?

**Comment by davidfowl** [comment]
Exposing a ValueTask would be sufficient but there’s already a bug for that (I think we had the discussion about why it was a callback instead of a task in another issue).

**Comment by garethwarry** [comment]
I agree. The way it's written right now is a bit confusing. It leads me to ask why OnComplete* runs asynchronously as a callback. Why doesn't it run in serial with the callee of the Complete method as an action like most other portions of the .NET Framework? I mean, if people want to spawn a new task for processing that OnComplete* "callback", then they can just add that to their passed action. All other callbacks I've seen in .NET have been very purposefully named for such functionality, like "BeginReceive" and "BeginSend" for sockets, "BeginRead" and "BeginWrite" for all streams, etc. If Completion was made a property, then I think "public virtual Task Completion" would be alright (like how System.Threading.Channels does it). Or make it an action and give that control to the implementer like how SocketAsyncSocketArgs does it with "public event EventHandler<T> Completed".

**Comment by davidfowl** [comment]
There’s a concept of a scheduler and the default scheduler is the thread pool scheduler. Callbacks are always scheduled on the configured scheduler. 

In hindsight, this should have been a value task since the implementation would have been pluggable. Events lack the ability to avoid closure capture when you want to pass state in so that’s why we have a method with a state parameter. It’s a pretty common pattern in most callback based APIs

**Comment by garethwarry** [comment]
Thanks for the clarification.

**Comment by davidfowl** [comment]
Prior art for the state parameter:

https://github.com/dotnet/corefx/blob/6a0ecec8279504c03433f689e4a7cf6cd63da27c/src/Common/src/CoreLib/System/Threading/CancellationToken.cs#L192

**Comment by davidfowl** [comment]
We'll consider this for a future milestone, we don't have a good way to signal that something bad happened. We'll consider using https://github.com/dotnet/corefx/issues/31159
